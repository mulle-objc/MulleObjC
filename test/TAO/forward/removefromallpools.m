#define TEST_TAO_STRATEGY   MulleObjCTAOCallerRemovesFromAllPools

#include "TAOStrategyTest.inc"


int   main( int argc, const char * argv[])
{
   NSThread    *thread;
   Foo         *obj;

#ifdef __MULLE_OBJC__
   if( mulle_objc_global_check_universe( __MULLE_OBJC_UNIVERSENAME__) != mulle_objc_universe_is_ok)
      return( 1);
#endif

   [NSThread mulleSetMainThreadWaitsAtExit:YES];
   [[NSThread currentThread] mulleSetNameUTF8String:"#1"];

   @autoreleasepool
   {
      obj = [Foo object];
      mulle_printf( "%s - 1: %s %td\n", __THREAD_NAME__, __PRETTY_FUNCTION__, [obj retainCount]);

      @autoreleasepool
      {
         thread = [NSThread alloc];
         [thread mulleSetNameUTF8String:"#2"];
         thread = [thread mulleInitWithObjectFunction:thread_function
                                               object:obj];
         thread = [thread autorelease];

         mulle_printf( "%s - 2: %s %td\n", __THREAD_NAME__, __PRETTY_FUNCTION__, [obj retainCount]);
         [thread start];
         mulle_relativetime_sleep( 0.5);  // wait for thread before autoreleasepool clean up for nicer log

         // can't access foo anymore here
      }
   }
   return( 0);
}
