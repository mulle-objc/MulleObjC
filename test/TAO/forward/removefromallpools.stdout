#1 - a<: -[Foo autorelease] 1
#1 - 1: int main(int, const char **) 1
#1 - +<: -[Foo retain] 2 -> 3
#1 - a<: -[NSThread(Debug) autorelease] 1
#1 - 2: int main(int, const char **) 2
#1 - +<: -[NSThread(Debug) retain] 2 -> 3
#1 - q>: -[Foo mulleRelinquishAccess] 2
#1 - +<: -[Foo retain] 3 -> 4
#1 - q<: -[Foo mulleRelinquishAccess] 2
#2 - g>: -[Foo mulleGainAccess] 2
#2 - a<: -[Foo autorelease] 2
#2 - g<: -[Foo mulleGainAccess] 2
#2 - 3*: thread_function 2
#2 - f>: -[NSThread(Debug) finalize] 2
#2 - a<: -[Foo autorelease] 2
#2 - f<: -[NSThread(Debug) finalize] 2
#2 - ->: -[Foo release] 2 -> 1
#2 - ->: -[Foo release] 1 -> 0
#2 - d>: -[Foo dealloc] 0
#? - ->: -[NSThread(Debug) release] 2 -> 1
#1 - ->: -[NSThread(Debug) release] 1 -> 0
#1 - d>: -[NSThread(Debug) dealloc] 0
#1 - f>: -[NSThread(Debug) finalize] 1
#1 - f<: -[NSThread(Debug) finalize] 1
#? - ->: -[NSThread(Debug) release] 1 -> 0
#? - d>: -[NSThread(Debug) dealloc] 0
